Iterate over the classifier choices per symbol

chargeflux opened the pull request:
I added a function that shows how the iterator would be implemented. It starts from line 117 because it shares code from the `recognize` function. It is not "complete" but it is in a workable state and represents how I approached it.

The way the function works is it gets an image, Tesseract extracts all the symbols from the image and for each symbol, the function gets all possible classifier choices for each symbol with CI. This results in a 3D array: an array of array of symbols where each symbol has an array of possible classifier choices + CI.

I am looking for any feedback on whether you think this is how the function should be implemented.

There are a few things to keep in mind. This implementation is "symbol"-specific (aka character-based) and based on the API, while the "PageIteratorLevel" can be symbol, line, paragraph or block, the use of the ChoiceIterator in line 133 below is symbol-only. That means Tesseract won't give confidence intervals for an entire line, which is understandable.

I don't think it make sense to have a ChoiceIterator and a 3D array for any other level other than symbol. If there are multiple symbols in a picture, then yes, you would want to get each symbol that Tesseract recognizes and get all classifier choices for each symbol, hence the 3D array.

It might be worthwhile to support for the PageIteratorLevels as a separate function from the symbol-specific implementation.

What are your thoughts?

chargeflux committed:
Add iterator for symbol + confidence interval

chargeflux committed:
Add documentation explaining how symbol iterator works

chargeflux committed:
Implement 3D array for classification possibilities for each symbol in a multi-symbol query

scott0123 commented:
Hi chargeflux, I've checked out your branch and verified for myself that what you have added was indeed working as you have claimed.

A couple of notes on my end:
- The function name is rather misleading, `getIterator` implies that we will return the iterator to the caller. Instead of letting the caller deal with the iterator, what we have done here is essentially processed the iterator for the user and compiled the results for them. I would recommend changing the name and adding a small block of documentation right above its declaration in the header file that explains the functionality and return values in detail.
- To address the "support for the PageIteratorLevels as a separate function" request, naively we might want to just return the different types of iterators directly to the caller; however, this isn't actually viable since it wouldn't be a valid return type for a standard Obj-C caller. Additionally, doing this might defeat the purpose of having the wrapper in the first place, since it is supposed to act as a layer of abstraction between the raw API and the user, but the iterator isn't really usable directly by the user without in-depth knowledge of how it is supposed to be used. Perhaps, the only real way to achieve what you want is by taking the iterator level that the user wishes to access and performing all the actions that they might want to perform for them, and returning the results. This would be similar to the current implementation but for different iterator levels.
- The length/depth of the CI choices for a symbol is not limited by anything on our end. Since this might be a rather long list, and the user might be iterating through a large text file with many symbols, it may be wise to cap it at a reasonable depth; say, 5.
- The return value of the function call does not need to be mutable any more. Consider returning the immutable version instead.
- The confidence level is returned by the `ci.Confidence()` call is actually a percentage (e.g. 98.3). Maybe we want to convert that to a value between 0 and 1.
- You have included a debugging function on line 108. Consider removing that when committing.

In general, You've done a great job on implementing this function. Congrats!
Please update the points I have listed and I'd be more than happy to merge this PR. If you think they still need discussion, I can follow up on that too.

Thanks for contributing and feel free to submit more issues/feature requests if you need them.

chargeflux committed:
Rename function and add necessary documentation in header and improve result value

Result value is now a NSArray instead of NSMutableArray and confidence interval is 0-1

chargeflux commented:
I have renamed the function to `getClassiferChoicesPerSymbol` and added documentation in the header file with an example that I came up with. I also added a comment about why one would need this function (line 36-38 in header file) with how Tesseract incorrectly classified the symbol as O but the next best choice is the correct answer ("C"). I thought it might be useful.

I will limit the scope of this PR to just the symbol iterator level and getting classifier choices per symbol. The rest of the iterator levels will require a separate PR because the assumed purpose for those are different from this current function. For example, as far as I can tell, you can't use the `choiceIterator` for those levels since it is per-symbol and if the `resultIterator` uses a level higher than `symbol`, then it wouldn't work. I haven't really delved in on whether I am right with this example though.

I also changed the confidence level to a value of 0 and 1 and returned an immutable NSArray. 

Regarding depth, I am not sure if we should artificially limit it. I think the user should handle the logic on what to do with the classifier choices per symbol since perhaps the user may have specific confidence interval cutoffs. Additionally, for languages like Chinese and Japanese, due to the sheer amount of characters, if we limit to the top 5 values for example, it might not be enough to get the supposed 'right' answer in the returned array. 

Perhaps there might be memory or performance concerns but I don't think the user would be using this function on a whole paragraph at once. Even if the user wanted to do it on a paragraph, they would probably split up the paragraph and call this function per line so it is manageable and more granular.

What do you think?

scott0123 commented:
Hi chargeflux,

I'm satisfied with your new changes and will be merging now.
Closing issue #1 as fixed.

chargeflux commented:
Thanks for the feedback and accepting my PR! I look forward to contributing to this repo again!

scott0123 commented:
I noticed that the "contributors" count is still 1 even well after I've merged your commits. @chargeflux 

If maintaining a polished GitHub profile matters to you, I would recommend linking an email to your GitHub account, and setting up that same email as the "user" when making commits, so that GitHub can recognize the commit as yours and list you as a contributor.
Right now, the email I see in the commit is <20979457+chargeflux@users.noreply.github.com>

Although I'd recommend not reverting those commits just to change that one thing, it might be worth it to linking emails for future contributions.

chargeflux commented:
If you click on the Contributors tab, I do show up as a second "contributor" on that page. It seems that the contributor count in the tab has not updated yet. I will wait about a day and see if it refreshes.

chargeflux commented:
@scott0123 I was using this function and I noticed there is actually a bug with my implementation. `symbol_result` (which holds all classifier choices for the symbol) is outside of the `resultIterator`’s do loop so it doesn’t get cleared on the next symbol iteration. Line 128 needs to switch with line 129. I can submit a quick fix or you can fix it yourself.

Not sure why I didn’t catch that earlier but I apologize for the error!

scott0123 commented:
Good catch @chargeflux . Fixed.
I should have tested it more too, so the fault is on me as well.
